turbo/snapshotsync/freezeblocks/block_reader.go

```diff
@@ -1250,6 +1253,53 @@ func (r *BlockReader) EventsByBlock(ctx context.Context, tx kv.Tx, hash common.H
 	return result, nil
 }
 
+func (r *BlockReader) EventsById(from uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error) {
+	view := r.borSn.View()
+	defer view.Close()
+
+	segments := view.Events()
+	var buf []byte
+	var result []*heimdall.EventRecordWithTime
+	stateContract := bor.GenesisContractStateReceiverABI()
+
+	for _, sn := range segments {
+		idxBorTxnHash := sn.Index()
+
+		if idxBorTxnHash == nil {
+			continue
+		}
+		if idxBorTxnHash.KeyCount() == 0 {
+			continue
+		}
+
+		offset := idxBorTxnHash.OrdinalLookup(0)
+		gg := sn.MakeGetter()
+		gg.Reset(offset)
+		for gg.HasNext() {
+			buf, _ = gg.Next(buf[:0])
+
+			raw := rlp.RawValue(common.Copy(buf[length.Hash+length.BlockNum+8:]))
+			event := heimdall.UnpackEventRecordWithTime(stateContract, raw)
+			if event.ID < from {
+				continue
+			}
+
+			result = append(result, event)
+
+			if event.Time.After(to) {
+				// we return an extra record to signify that we exit because of time and not limit
```

Comment:
ah this feels confusing - the caller is asking for event from "fromId" to "to time" but we are giving them events with last event after "to time" - I'd suggest we dont do this

instead the caller should always get events before "to" and in the edge case where the caller gets a response for which len(events)==limit then the caller should call EventsById again but with from=events[len(events)-1] - this is similar to a paging api. If there are no event ids before "to" left on the subsequent calls then the func will return empty list